packages/tasit-action/src/TasitAction.test.js

```diff
@@ -84,109 +77,292 @@ describe("Contract", () => {
+    afterEach("waiting for message/tx confirmation", async () => {
+      if (subscription) {
+        await subscription.waitForMessage();
+        subscription.removeAllListeners();
+      }
+    });
 
-    const onMessage = async message => {
-      // message.data = Contents of the message.
-      const { data } = message;
-      const { confirmations } = data;
+    it("should throw when subscribing with invalid trigger", async () => {
+      subscription = simpleStorage.setValue("hello world");
+      return expect(subscription.on("invalid", () => {})).to.be.rejected;
+    });
 
-      if (confirmations === 7) {
-        subscription.removeAllListeners();
+    it("should throw when subscribing without callback", async () => {
+      subscription = simpleStorage.setValue("hello world");
+      return expect(subscription.on("confirmation")).to.be.rejected;
+    });
+
+    it("should call a write contract method (send tx) - imediate subscription", async () => {
+      var rand = Math.floor(Math.random() * Math.floor(1000)).toString();
+      subscription = simpleStorage.setValue(rand);
+      const fakeFn = sinon.fake();
+
+      const onMessage = async message => {
+        const { data } = message;
+        const { confirmations } = data;
+
+        if (confirmations >= 7) {
+          fakeFn();
+
+          subscription.removeAllListeners();
+
+          const value = await simpleStorage.getValue();
+          expect(value).to.equal(rand);
+
+          // FIXME: UnhandledPromiseRejectionWarning
+          //expect(1).to.equal(2);
+        }
+      };
+
+      await subscription.on("confirmation", onMessage);
+
+      await mineBlocks(simpleStorage.getProvider(), 15);
+
+      expect(fakeFn.called).to.be.true;
+    });
+
+    it("should call a write contract method (send tx) - late subscription", async () => {
+      var rand = Math.floor(Math.random() * Math.floor(1000)).toString();
+      subscription = simpleStorage.setValue(rand);
+      const fakeFn = sinon.fake();
+
+      await mineBlocks(simpleStorage.getProvider(), 15);
+
+      const onMessage = async message => {
+        const { data } = message;
+        const { confirmations } = data;
+
+        if (confirmations >= 7) {
+          fakeFn();
+
+          subscription.removeAllListeners();
+
+          const value = await simpleStorage.getValue();
+          expect(value).to.equal(rand);
+        }
+      };
+
+      await subscription.on("confirmation", onMessage);
+
+      await mineBlocks(simpleStorage.getProvider(), 20);
+
+      expect(fakeFn.called).to.be.true;
+    });
+
+    it("should remove listener after timeout", async () => {
+      subscription = simpleStorage.setValue("hello world");
+      const confirmationFn = sinon.fake();
+      const errorFn = sinon.fake();
+
+      const foreverCallback = async message => {
+        confirmationFn();
+      };
+
+      await subscription.on("confirmation", foreverCallback);
+
+      await subscription.on("error", err => {
```

Comment:
I forget if we have it this way in the source code, but if there's an error after a certain amount of time, we should reset the timer to zero after each additional confirmation

Reply:
Ok, review `time-out` logic is on my radar because I think that the way it's working now isn't good enough. As it's now, all alive listeners are throwing error event after the timeout. Could you specify for me more detailed when the timeout event should be thrown?

Reply:
Ah yeah, we wouldn't them all to throw an error event

The goal should be for just one error event to come in when this happens.

I think the time when it should be thrown is if

`current time - last confirmation block time > CONFIRMATION_TIMEOUT`

Reply:
Resolving this!